security/pkg/nodeagent/secretfetcher/secretfetcher.go

```diff
@@ -114,6 +121,54 @@ func (sf *SecretFetcher) Run(ch chan struct{}) {
 	cache.WaitForCacheSync(ch, sf.scrtController.HasSynced)
 }
 
+// Generate implements the XDS Generate method for SDS.
+// We also allow returning the secrets over the main XDS connection,
+// for efficiency.
+// This method only allows returning secret for authenticated requests.
+func (sf *SecretFetcher) Generate(proxy *istiomodel.Proxy, push *istiomodel.PushContext,
+	w *istiomodel.WatchedResource, updates istiomodel.XdsUpdates) istiomodel.Resources {
+	// 1. Get the namespace of the caller
+	// TODO: add an error to Generate (for non-empty)
+
+	// This depends on in-progress PR to VERIFY the namespace.
+
+	// 2. If istio-system - ok, can read secrets in istio-system
+	// This istiod can either be the 'master' in istio-system, or
+	// an replica of istiod running in a namespace with reduced
+	// RBAC to only namespace secret access. The second is equivalent
+	// with running the gateway with secret access, but secret permissions
+	// are granted to a separate istiod pod.
+	if proxy.Metadata.Namespace != sf.secretNamespace {
+		// First iteration: we only support istio-system ( or ingress + istiod in same namespace )
+		// If Ingress is in a different namespace - either use the old code or
+		// run a local istiod for SDS access.
+		return nil
+	}
+
```

Comment:
another thing to keep in mind - we are essentially bypassing k8s by saying any proxy in a namespace can read any secret. Doesn't k8s rbac have more controls to say only certain service accounts can read a particular secret? if so, shouldn't we be checking the same rbac permissions before returning the secret? 

If we dont support the following scenario, we are essentially adding CVE  :P.

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: default
  name: secret-reader
rules:
- apiGroups: [""]
  resources: ["secrets"]
  resourceNames: ["secret-for-service-1"]
  verbs: ["get"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: read-secrets
  namespace: default
subjects:
- kind: ServiceAccount
  name: svc1-service-account
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: Role
  name: secret-reader
  apiGroup: rbac.authorization.k8s.io
```

In the above case, only pods with svc1-service-account should be able to touch the `secret-for-service-1` and no body else. This applies to all, even the istio gateways as we do not know what the user's final install is.

Reply:
I completely agree - and the initial goal for this PR is to just handle istio-system ( or the namespace where istiod is running and where we grant namespace-only secret read). 

Accessing secrets in other namespaces, and by other gateways is quite controversial, but I think everyone agrees that it'll need to have an access control equivalent with RBAC. 

Note that RBAC controls access to k8s CRDs, but it is not very fine grained. Istiod can apply additional restrictions,
like requiring some label/annotation on the secret that point to a gateway, or explicit configs. 

The other intent of this is to support Vault or other secret stores - using a SDS adapter. Same requirements of having a policy check based on the service account running the gateway. But I don't think we can require everyone to use RBAC - nor is RBAC granular enough.

In any case - this PR is not attempting to do 'all namespaces'. Once we start having some Vault or other integrations - we can try to unify a policy or best practices.